### ISSUE TITLE
audio xruns on track load

### ISSUE THREAD

**Issue by Holzhaus** [Holzhaus/mixxx-gh-issue-migration#334]
how to reproduce the bug :
- launch jack with correct settings (after lots of sleepless nights, i found them...)
- run mixxx (audio configured with jack, output to 1-2, headphones to 5-6)
- load a track on deck1
- play it
- load a track another track on deck 2
at this time appear small/big glitches
- load yet another track on deck 2
glitches again
etc...

i'm on Karmic on 2.6.31-19-generic
on a Vaio laptop Intel(R) Pentium(R) 4 CPU 2.80GHz
graphic card radeon mobility 9200
audio card TerraTec Electronic GmbH Aureon 5.1 MkII
desktop environment LXDE/Openbox

actually, there are glitches also on 1.7.2 version but much less (a small one when the waveform is being displayed i found).

**Comment by Holzhaus** [comment]
(by rryan)
Hey samotham,

When you say glitch, do you mean audio glitch or graphical glitch?

Thanks,
RJ

**Comment by Holzhaus** [comment]
(by samotham)
I meant audio glitch

**Comment by Holzhaus** [comment]
(by rryan)
This is a known issue though I don't think we have a bug for it. I have a lot of work done in the lp:mixxx/features_sqlite branch which drastically improve this. These fixes will be incorporated into Mixxx 1.8.0 beta2 so please give that a try when it comes out. For now, you can increase your latency to eliminate audio xruns on track load.

Thanks very much for the bug report,
RJ Ryan

**Comment by Holzhaus** [comment]
(by samotham)
ok good

then to be more precise, for info these xruns are not considered as real xruns by jack

**Comment by Holzhaus** [comment]
(by samotham)
for information as it might help
back to 1.7.2, I still get some minor kind of xruns when loading a new track. Precisely, it appears just before the BPM is being displayed and waveform becoming fluid.

In the meantime, this week-end I checked on Windows XP, I don't get any xruns or glitch (using ASIO4all drivers on the same soundcard).

BTW shall I open a new bug report about this ?

**Comment by Holzhaus** [comment]
(by sentenzux)
Hi guys,

I'm getting the same behavior with karmic on a desktop system with a Q6600, so I guess performance should not be a problem... Setup is based on a edirol FA101 (firewire) and an external mixer, vinyl control is not used. Ardour is used for recording the mixer output via an input of the FA101.

If I first load my tracks a first time, reloading do not produce the same effect. Jack timeout is set to 2 sec (because otherwise mixxx crashes), and I do get around 1-2 sec of blank in my live output when I get this problem, however my recording by ardour only has a really small glitch and 1 xrun displayed.

**Comment by Holzhaus** [comment]
(by gamegod)
This situation should be significantly improved in trunk. Features_sqlite has been merged into trunk for some time (which brings in a slew of library performance enhancements), and we also now use memory mapped I/O for MP3 playback, which should also speed things up.

**Comment by Holzhaus** [comment]
(by roland-robelix)
Is this already in trunk?

I get audio glitches almost always when loading a song with rev 2421.
Running on Gentoo, kernel 2.6.32-gentoo-r7, plain ALSA, no jack, no pulseaudio.
T43, 2G RAM, 2GHz Pentium M, Mobility Radeon X300, NI A2DJ

1.7.2 runs fine with 21ms latency, 1.8 produces glitches even with 170ms.

**Comment by Holzhaus** [comment]
(by gamegod)
Yeah, this is in trunk. I recently made changes to SoundSourceMP3
which could have possibly done some bad things, but I got Owen to
review them to make sure they were OK so I don't think I've done
anything super dumb there (though it's still very possible).

Do you get xruns when you load OGG or FLAC files?

RJ, you were able to get xruns on load before you added the caching
stuff, right? Can you test on your rig and see we've regressed on this
recently?

Thanks,
Albert

**Comment by Holzhaus** [comment]
(by roland-robelix)
It happens when loading mp3 or ogg, not with flac.
The type of the playing file has no influence.

**Comment by Holzhaus** [comment]
(by bkgood)
Robelix, have you tried lowering your latency? I'm toying with a bug similar to this one (may be the same one), and I only seem to get these glitches on track load if my latency is >=10 ms.

Could you try 1 or 5 ms to see if the glitches still occur?

**Comment by Holzhaus** [comment]
(by pegasus-renegadetech)
I still see this on my slowww Celeron D 2.53GHz. It is worse at higher latencies, but still detectable at 10ms.

**Comment by Holzhaus** [comment]
(by d00guan)
I've done some testing now, since I find it to be the most annoying thing with Mixxx right now. Otherwise 1.8 it looks promising!

I see it on all file types, though much more frequently for mp3 than for ogg, flac or wav. Then my q6 ogg files are around half the size of my mp3's which are mainly 320kbit. My system is a 1.8GHz Core2Duo (7100), X3100 laptop with 4gig ram running ubuntustudio 10.04 64bit. Reducing latency from around 10ms to 5ms gives more xruns. Mixxx version is trunk r2432.

**Comment by Holzhaus** [comment]
(by bkgood)
Ok I ought to throw my two cents in here because this seems like it should block 1.8 as it really makes Mixxx unusable for serious mixing (loading a new song kills your master out, if only for half a second).

I may be totally wrong here, but this seems to happen because ReadAheadManager (or CachingReader) insists on reading a handful (8 or so I think) of chunks from the file before it starts feeding to EngineBuffer, and it seems to block the rest of the audio callback thread until these 8 chunks are read. The ideal behavior would be to either just read what it needs and then fill in the rest as possible with another thread, but that appears to be what's happening now (or that's how I'm reading the code), so maybe the ideal behavior is to just keep returning buffers of zeros until the 8-or-whatever chunks have been acquired and playback can begin.

I haven't found where this blocks but that's about all I can figure is happening.

FWIW I first noticed this bug testing different FLAC files with my flac sound source, I noticed it starts skipping when the file is encoded at -3 or better (-3 and higher corresponds to a 4092 frame block size, so it's likely more difficult to decode), this bothered me for a while before I realized there was no way I could really make my code noticeably faster so I tested the same set of files with SSsndfile and found the same skips (at the same encoder levels) happen there too. This happens to me occasionally when playing other file types too (mp3 and m4a specifically); I think it's reasonable that Mixxx would skip if a soundsource couldn't decode one chunk of data in a callback period, but asking it to decode 8 or 10 of them in that short of time is quite a lot.

**Comment by Holzhaus** [comment]
(by roland-robelix)
Just tried the beta2 - and I also tried with latency set to 1,2 and 5ms

Since I get the "normal too low latency noises" all the time it's hard to say if there are additional glitches while loading a song.

**Comment by Holzhaus** [comment]
(by roland-robelix)
Think I've got that beast ;)

I built the latest portaudio snapshot (2010-07-30)  - and it runs just fine.

**Comment by Holzhaus** [comment]
(by bkgood)
On Saturday 31 July 2010 15:07:59 Robelix wrote:
> Think I've got that beast ;)
> 
> I built the latest portaudio snapshot (2010-07-30)  - and it runs just
> fine.

PortAudio has gone some optimization lately, so glad to know it fixed it for 
you. :) However, this definitely doesn't resolve whatever's going on in the 
mixxx codebase that was causing this to begin with (i.e. loading a song seems 
capable of blocking the audio callback thread).

Bill

**Comment by Holzhaus** [comment]
(by d00guan)
What Bill states sounds like a likely cause for the behaviour while loading tracks. Something seem to block for an instance. I bet this is the same cause that make jackd kick Mixxx if one don't use ridiculous values for timeout.

Is there someone with insight in this code that can confirm or deny what Bill states? For me this is a blocker, since I don't want these kind of breakups when playing live.

**Comment by Holzhaus** [comment]
(by rryan)
Hi Anders, Bill, all,

There isn't really any code in the engine that will block based on an action
taken in the UI. I could be wrong, but I think this problem is mostly just a
combination of too-low latency setting and the various debug statements we
have that print to the console on track load.

I just did an experiment where I turned off the EQs and the waveform
summary, set one track playing, and then spam-loaded every track in my
library to the second player. This ensures that the analyzer thread spends
100% of its time processing tracks (load from disk, calculate waveform,
calculate BPM, etc).

The result was that audio xruns still occured at 5ms latency.

I took one more step and converted the ControlObject sync step to happen
outside of the audio callback thread. The result was the same, xruns still
occured on track load when I spam-loaded tracks to the player.

I then piped the output of Mixxx to /dev/null and retried. The result was
the xruns occurred much less than before, very rarely. I even rate limited
my CPU from 2.4GHz to 800MHz and xruns very very rarely occured.

I've reviewed the engine code again and there really isn't anything that
will obviously block. It's possible theres something in there, but it's
hiding very well.

I think that moving ControlObject::sync() out of SoundManager and into an
EngineWorker is a good move overall, because nasty people (e.g. MIDI Script)
could connect valueChanged or valueChangedFromEngine signals from a sync()
to do some long running work or something that could block the engine
thread, so a sync() call should not be in the engine callback.

RJ

On Mon, Aug 9, 2010 at 3:12 AM, Anders Gunnarsson <<email address hidden>
> wrote:

> What Bill states sounds like a likely cause for the behaviour while
> loading tracks. Something seem to block for an instance. I bet this is
> the same cause that make jackd kick Mixxx if one don't use ridiculous
> values for timeout.
>
> Is there someone with insight in this code that can confirm or deny what
> Bill states? For me this is a blocker, since I don't want these kind of
> breakups when playing live.
>
> --
> audio xruns on track load (1.8.0beta1)
> https://bugs.launchpad.net/bugs/529614
> You received this bug notification because you are a bug assignee.
>

**Comment by Holzhaus** [comment]
(by d00guan)
Bummer.

The thing regarding debug messages sounds interesting though. Would it be easy to disable log of these messages by moving them to a log level other than default?

I get xruns even with Performance mode 1,8GHz on both my cores on a Core2Duo with ~10ms latency, and Mixxx as the only load on the system. Although I use jack on a realtime kernel, and have not been to pedant with what runs on my system, I think that it should be possible to get it free of xruns on track load.

If this was fixed Mixxx would really be a professional grade DJ software. Probably the best performing I know of.

**Comment by Holzhaus** [comment]
(by pwhelan)
Have you tried disabling the BPM detection and also getting output to pipe to /dev/null?

I'm thinking that if it's caused by the BPM detection code we could limit it's CPU usage dynamically. We just incur a small sleep if a process call to the analyzer uses too much CPU.

**Comment by Holzhaus** [comment]
(by pwhelan)
Here is a patch for the 1.8 release that should throttle the CPU for the Analyzer Queue. If someone who is experiencing the XRUN problem could try it out and tell me if it fixes the problem it would be very helpful.

The patch has a few quirks but should be stable. Ocassionally it will over throttle. If this does fix the problem I'll go ahead with the cosmetics and such to get this into the final 1.8 release.

**Comment by Holzhaus** [comment]
(by daniel-64studio)
I noticed something similar using 1.8.0-beta2 when loading FLAC or WAV files in ALSA mode on 64 Studio 3.3 alpha (Karmic based with an RT kernel). To reproduce, get a track rolling in one player, then drag and drop another track from the library into the other player. There is a very short audio glitch in the playing track, but it's noticeable.

At first I thought this was due to the high default latency setting of 64ms - I dropped it to 10ms and the problem stopped temporarily. But when I restarted Mixxx, still set to 10ms, the problem came back.

BPM detection could be the cause, because I'm testing with a freshly imported library of audio files. Can I enable batch BPM detection, waveform generation etc. at library import time, rather than when I load the track into the player for the first time? OK, import would take longer, but then I'm not likely to be doing a mass file import in the middle of a live set.

I am running Mixxx on this HP ze2000 laptop as the sole application (no gdm, mixxx started from an .xinitrc file) so there is no desktop or window manager to interfere. The soundcard is a basic internal AC97 one at the moment. If I can't track down the cause, I will try Philip's patch above, and let you know if that fixes it.

Cheers!

Daniel

**Comment by Holzhaus** [comment]
(by gamegod)
Here's my 2 cents:

1) We can't sleep-limit the Analyser thread, just on principle. If you
have a 10 GHz processor, we're not going to arbitrarily slow down any
of our background threads on you. It's like the opposite of old DOS
games running too fast on modern PCs, but it's the same type of
thinking. There's a reason your OS has a scheduler.

2) The only people that are complaining here are Linux users.

3) Do you guys have the same xruns if they disable the waveform views?
(Change the view to "Simple" or "Off" or whatever we call it now in
the preferences.)

4) You can do batch BPM analysis via the "Analyze" tab in 1.8. We
don't currently do waveform generation in advance because we're not
entirely sure we want to save that to disk. (eg. if the waveform
pixmap is 100 kb, then we'd need 1 GB of your disk space to cache the
waveform for every song in your 10,000 song library.)

5) Your OS has a scheduler, let it work. If you have an RT thread, I'm
very surprised that you could preempt it via some GUI action. The call
graph that takes place when a song is loaded into a player is fairly
complicated, and it would definitely help to have a diagram of this.
If I have time to review the code again, I'll try to draw one.

Thanks,
Albert

**Comment by Holzhaus** [comment]
(by pwhelan)
Here's my answer

1) The OS Scheduler under linux is highly uncooperative. Getting it to work requires at the very least chaning /etc/security/limits.conf. Should we see if our package maintaners would be ok with rolling out a /etc/security/limits.conf or limit.d file? This might be the best alternative. Hopefully it is sufficient.

2) I'm a linux user (grin)

3) I'd like to hear the answer to this one from some of the users having this problem.

4) I know, batch analysis is awesome and hopefully it'll get awesomer.

5) I currently have no problems with my OS scheduler since I'm running the CK kernel (from the ppa) and I added permissions to nice to -5 and change the rtprio to 90 to the audio group in my /etc/security/limits.conf file.

**Comment by Holzhaus** [comment]
(by gamegod)
Good points, Phil. We should find out if these users have tweaked
limits.conf files or not.

Daniel: Does 64 Studio ship with a limits.conf file that allows
applications to control the priorities of their own threads?
eg.  @audio - rtprio 90

We've found experimentally that by default, the Linux kernel does not
allow applications to lower the priority of their own threads. This is
terrible for Mixxx because it means our realtime audio thread might
actually be running with the same priority as our background worker
threads, and we've found that that can be a significant source of
xruns and performance degradation. Tweaking limits.conf allows Mixxx
to lower the priority of it's own threads, and provides a crucial
performance boost that might even put Linux on par with Vista+ and OS
X.

If anyone else hasn't tried tweaking this file, please follow the
instructions under "Real-Time Support" here:
https://help.ubuntu.com/community/UbuntuStudioPreparation
(Remember to add your user to the "audio" group too!)

Thanks,
Albert

**Comment by Holzhaus** [comment]
(by rryan)
Also, on Ubuntu proper, you need to either add yourself to the audio group, or make the limits.conf adjustments be for your username and not @audio. This is because on Ubuntu Desktop, pulse is the only member of the audio group.

**Comment by Holzhaus** [comment]
(by d00guan)
I've made some more testing. My setup is a C2D running UbuntuStudio 10.04 with rt-kernel and Jack.

Jack is started with:
jackd -R -t5000 -dalsa -dhw:1 -r44100 -p256 -n2

UbuntuStudio adds following to /etc/security/limits.d/audio.conf when jackd is installed:
@audio   -  rtprio     99
@audio   -  memlock    unlimited
#@audio   -  nice      -19

I also tried to uncomment the last row, but don't think it changed performance.

With this setup I get xruns almost on every track load. Even though I have both kernels in Performance mode. If I disable waveforms I still get xruns but they occur less frequently.

After this I tested ck preemptive kernel with much better result. Thanks for the tip Phil! 
I can still get xruns to occur if I spam load tracks though, and may get an xrun on normal track load, but not to often. I attached a log of an xrun and of what happens during spam load described below. 


While spam loading tracks I noted that tracks queued for analysis aren't removed even if a new track is loaded. The result is that every track is processed in sequence until the currently loaded track is processed. I know it isn't a normal use case, but it leads to an unnecessary delay until waveform is showed if a user load wrong track before loading the correct one.

**Comment by Holzhaus** [comment]
(by rryan)
Hi Anders,

Thanks for the data. Did you pipe the output of Mixxx to a file while
running it or was it outputting to stdout?

RJ

**Comment by Holzhaus** [comment]
(by roland-robelix)
I did a few tests with the beta2 on Gentoo x86 and the older portaudio from portage:

Switched off the Waveforms: no change.

Switched off BPM detection: no change.

Rea-time: Tried with the settings from the Ubuntu-page mentioned above and from http://www.gentoo.org/proj/en/desktop/sound/realtime.xml - no change. (BTW: How do I check if RT is really working?)

Reniced mixxx to -15: no change.

Diffrent latency settings change how the glitches sound but not the fact that there are some.

**Comment by Holzhaus** [comment]
(by d00guan)
Hi RJ,

It's console output, so I guess stdout. I skipped the initial library scan and init log.

**Comment by Holzhaus** [comment]
(by rryan)
Hey Robelix, 

Interesting, earlier on you said that updating to portaudio trunk helped. Is that still the case?

RJ

**Comment by Holzhaus** [comment]
(by rryan)
Just got done messing around a little bit. 

I have a Core 2 Duo 2.5GHz laptop (Lenovo T400) with 2gb of RAM. I wrote a patch that completely turns off any debug prints that happen as a result of loading a track. As a result, while running tons of other apps on plain old Ubuntu Desktop 10.04 with the limits.conf settings correctly set and SpeedStep ENABLED locking my CPU to 800MHz Mixxx running at 5ms latency through ALSA has 0 xruns. Not a single one. I spam-load my entire library and can't get it to xrun while a track is playing. (the spam-load pins the CPU at 100%, but still no xruns) I also tried using JACK connected through ALSA and still couldn't get an xrun. If I turn off this patch so Mixxx does debug output on track load again, I get xruns all over the place. 

So, robelix, Anders, can you try piping the output to /dev/null and see what happens? The patch I applied is here if you are able to compile Mixxx 1.8 from source.  http://pastebin.com/cd0vGzR5

**Comment by Holzhaus** [comment]
(by rryan)
Oh, I'm using the latest version of portaudio-trunk, which has some performance improvements, but I'm definitely able to get xruns with portaudio trunk if I enable all those print statements.

**Comment by Holzhaus** [comment]
(by rryan)
Actually, I just checked. Piping output to /dev/null still gives me xruns. If I run without those printlines enabled, no xruns. If you guys are able to compile from source, please apply my patch and test.

**Comment by Holzhaus** [comment]
(by roland-robelix)
Ryan, qith svn-portaudio it's far better but not perfect - xruns happen only once in about 100 songs.
For testing I installed an older version of portaudio.

And I wanted to try your patch - but I'm not able to appy it (tried with beta2):

patching file src/analyserbpm.cpp
Hunk #1 FAILED at 44.
Hunk #2 FAILED at 72.
2 out of 2 hunks FAILED -- saving rejects to file src/analyserbpm.cpp.rej
patching file src/analyserwaveform.cpp
Hunk #1 FAILED at 53.
Hunk #2 FAILED at 108.
2 out of 2 hunks FAILED -- saving rejects to file src/analyserwaveform.cpp.rej
patching file src/analyserwavesummary.cpp
Hunk #1 FAILED at 77.
1 out of 1 hunk FAILED -- saving rejects to file src/analyserwavesummary.cpp.rej
patching file src/library/dao/cuedao.cpp
Hunk #1 FAILED at 228.
Hunk #2 FAILED at 244.
2 out of 2 hunks FAILED -- saving rejects to file src/library/dao/cuedao.cpp.rej
patching file src/library/dao/trackdao.cpp
Hunk #1 FAILED at 101.
Hunk #2 FAILED at 316.
Hunk #3 FAILED at 343.
Hunk #4 FAILED at 489.
Hunk #5 FAILED at 556.
Hunk #6 FAILED at 719.
6 out of 6 hunks FAILED -- saving rejects to file src/library/dao/trackdao.cpp.rej
patching file src/soundsourcemp3.cpp
Hunk #1 FAILED at 26.
Hunk #2 FAILED at 37.
Hunk #3 FAILED at 139.
Hunk #4 FAILED at 147.
Hunk #5 FAILED at 524.
Hunk #6 FAILED at 600.
Hunk #7 FAILED at 626.
Hunk #8 FAILED at 638.
Hunk #9 FAILED at 688.
Hunk #10 FAILED at 715.
10 out of 10 hunks FAILED -- saving rejects to file src/soundsourcemp3.cpp.rej
patching file src/trackinfoobject.cpp
Hunk #1 FAILED at 273.
1 out of 1 hunk FAILED -- saving rejects to file src/trackinfoobject.cpp.rej
patching file src/waveform/waveformrenderer.cpp
patch unexpectedly ends in middle of line
Hunk #1 FAILED at 496.
1 out of 1 hunk FAILED -- saving rejects to file src/waveform/waveformrenderer.cpp.rej

**Comment by Holzhaus** [comment]
(by rryan)
You may need to check out the latest version from Bazaar. If you have bzr
installed, just do:

`bzr checkout lp:mixxx/1.8'

And that should get you a checkout of the latest 1.8 code. If you can
already build the beta2 code, that should work fine. The patch should apply
cleanly against that.

**Comment by Holzhaus** [comment]
(by roland-robelix)
same result with bzr version, the patch fails.

**Comment by Holzhaus** [comment]
(by gamegod)
FWIW the patch worked for me on latest 1.8, though:

alb@jupiter:~/Projects/mixxx-1.8$ patch -p0 < /home/alb/Downloads/cd0vGzR5.txt
(Stripping trailing CRs from patch.)
patching file mixxx/src/analyserbpm.cpp
(Stripping trailing CRs from patch.)
patching file mixxx/src/analyserwaveform.cpp
(Stripping trailing CRs from patch.)
patching file mixxx/src/analyserwavesummary.cpp
(Stripping trailing CRs from patch.)
patching file mixxx/src/library/dao/cuedao.cpp
(Stripping trailing CRs from patch.)
patching file mixxx/src/library/dao/trackdao.cpp
(Stripping trailing CRs from patch.)
patching file mixxx/src/soundsourcemp3.cpp
(Stripping trailing CRs from patch.)
patching file mixxx/src/trackinfoobject.cpp
(Stripping trailing CRs from patch.)
patching file mixxx/src/waveform/waveformrenderer.cpp
patch unexpectedly ends in middle of line
Hunk #1 succeeded at 496 with fuzz 1.

With RJ's patch, I don't get xruns on track loads, though I didn't do
great A-B testing here. I tested at 5 ms on a 3 GHz C2D (E8500) and
also with it clocked down the 2 GHz, which is the lowest it'll go.

(I still get the usual assortment of insane other Linux xruns during
regular playback, such as the ones caused by my wireless network
driver. Ubuntu 9.10 with stock kernel. )

Albert

**Comment by Holzhaus** [comment]
(by roland-robelix)
Tried again with a fresh checkout - no luck ;(

dings ~ # mkdir mixxx
dings ~ # cd mixxx/
dings mixxx # bzr checkout lp:mixxx/1.8 
You have not informed bzr of your Launchpad ID, and you must do this to
write to Launchpad or access private data.  See "bzr help launchpad-login".
dings mixxx # ls -lh 
total 4.0K
drwxr-xr-x 11 root root 4.0K Aug 23 12:13 1.8
dings mixxx # wget http://pastebin.com/download.php?i=cd0vGzR5
--2010-08-23 12:16:03--  http://pastebin.com/download.php?i=cd0vGzR5
Resolving pastebin.com (pastebin.com)... 173.236.86.29
Connecting to pastebin.com (pastebin.com)|173.236.86.29|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 11029 (11K) [application/download]
Saving to: `download.php?i=cd0vGzR5'

100%[================================================================================================================================>] 11,029      48.1K/s   in 0.2s    

2010-08-23 12:16:03 (48.1 KB/s) - `download.php?i=cd0vGzR5' saved [11029/11029]

dings mixxx # cd 1.8/
dings 1.8 # patch -p0 < ../download.php\?i\=cd0vGzR5 
patching file mixxx/src/analyserbpm.cpp
Hunk #1 FAILED at 44.
Hunk #2 FAILED at 72.
2 out of 2 hunks FAILED -- saving rejects to file mixxx/src/analyserbpm.cpp.rej
patching file mixxx/src/analyserwaveform.cpp
Hunk #1 FAILED at 53.
Hunk #2 FAILED at 108.
2 out of 2 hunks FAILED -- saving rejects to file mixxx/src/analyserwaveform.cpp.rej
patching file mixxx/src/analyserwavesummary.cpp
Hunk #1 FAILED at 77.
1 out of 1 hunk FAILED -- saving rejects to file mixxx/src/analyserwavesummary.cpp.rej
patching file mixxx/src/library/dao/cuedao.cpp
Hunk #1 FAILED at 228.
Hunk #2 FAILED at 244.
2 out of 2 hunks FAILED -- saving rejects to file mixxx/src/library/dao/cuedao.cpp.rej                                                                                    
patching file mixxx/src/library/dao/trackdao.cpp                                                                                                                          
Hunk #1 FAILED at 101.                                                                                                                                                    
Hunk #2 FAILED at 316.                                                                                                                                                    
Hunk #3 FAILED at 343.                                                                                                                                                    
Hunk #4 FAILED at 489.                                                                                                                                                    
Hunk #5 FAILED at 556.                                                                                                                                                    
Hunk #6 FAILED at 719.                                                                                                                                                    
6 out of 6 hunks FAILED -- saving rejects to file mixxx/src/library/dao/trackdao.cpp.rej                                                                                  
patching file mixxx/src/soundsourcemp3.cpp                                                                                                                                
Hunk #1 FAILED at 26.                                                                                                                                                     
Hunk #2 FAILED at 37.                                                                                                                                                     
Hunk #3 FAILED at 139.                                                                                                                                                    
Hunk #4 FAILED at 147.                                                                                                                                                    
Hunk #5 FAILED at 524.                                                                                                                                                    
Hunk #6 FAILED at 600.                                                                                                                                                    
Hunk #7 FAILED at 626.                                                                                                                                                    
Hunk #8 FAILED at 638.                                                                                                                                                    
Hunk #9 FAILED at 688.                                                                                                                                                    
Hunk #10 FAILED at 715.                                                                                                                                                   
10 out of 10 hunks FAILED -- saving rejects to file mixxx/src/soundsourcemp3.cpp.rej                                                                                      
patching file mixxx/src/trackinfoobject.cpp                                                                                                                               
Hunk #1 FAILED at 273.                                                                                                                                                    
1 out of 1 hunk FAILED -- saving rejects to file mixxx/src/trackinfoobject.cpp.rej                                                                                        
patching file mixxx/src/waveform/waveformrenderer.cpp                                                                                                                     
patch unexpectedly ends in middle of line                                                                                                                                 
Hunk #1 FAILED at 496.                                                                                                                                                    
1 out of 1 hunk FAILED -- saving rejects to file mixxx/src/waveform/waveformrenderer.cpp.rej                                                                              
dings 1.8 #

**Comment by Holzhaus** [comment]
(by d00guan)
Patching worked for me with patch -p0 < patch.diff in 1.8 folder.

For me it didn't do much improvement I'm sad to say. Maybe it's a tiny bit better, but I still get xruns often with rt-kernel, and can produce them with ck-preemptive kernel, but only if I load a few tracks rapidly after one another.

My setup is stable enough now, but the performance boost came mainly from using ck-preemptive instead of rt kernel from UbuntuStudio ppa.

**Comment by Holzhaus** [comment]
(by daniel-64studio)
Hi Albert, in reply to your comment #26, yes, the rtlimits are fully tweaked. I can run other CPU intensive JACK applications without xruns at the same buffer settings, so I don't think this is a system bug.

I no longer believe this is BPM detection related, because it happens even with previously analysed tracks. Subjectively, I'd say it's less noticeable when using JACK than when going direct to ALSA, but it's still there. 

I followed RJ's suggestion of piping output to /dev/null like this:

$ mixxx > /dev/null 2>&1

I'm not getting xruns, but I am still hearing the glitch when loading the second track. It doesn't matter if it's dragged and dropped or added with a right-click from the Library. So I'll try RJ's patch next.

Cheers!

Daniel

**Comment by Holzhaus** [comment]
(by daniel-64studio)
More info - I can't reproduce this problem on a stock Lucid system, even when running through Pulseaudio at 2ms. The original poster is running Karmic and I can definitely reproduce it there. So is this due to some bug in a Karmic version of a library and not Mixxx at all?

Anders, I think you have a different problem with your JACK performance. Which soundcard are you using?

**Comment by Holzhaus** [comment]
(by d00guan)
I'm using the card in my Hercules RMX. It's a generic usb sound device.

Is there any way to test the performance of a soundcard with alsa?

**Comment by Holzhaus** [comment]
(by daniel-64studio)
Update: I can reproduce this bug on Lucid by killing Pulseaudio, running jackd and connecting Mixxx to jackd - the glitch on track load comes back. Stop jackd, start Pulseaudio and connect Mixxx to that instead, the glitch goes away. Same if I stop Pulseaudio and connect direct to ALSA. 

However, if I install Mixxx 1.7.2 on the same system, the bug does not appear at high JACK buffer sizes such as 512 x 3. It does appear at 256 x 2 on the stock Ubuntu kernel.

**Comment by Holzhaus** [comment]
(by daniel-64studio)
Hi Anders, as you have a USB device, try:

jackd -R -t5000 -dalsa -dhw:1 -r44100 -p256 -n3

Some USB devices are just not happy with -n2. Going to -n3 will increase latency, of course. See also:

http://alsa.opensrc.org/index.php/Usb-audio#Tuning_USB_devices_for_minimal_latencies

There is a latency testing tool here: ftp://ftp.suse.com/pub/people/tiwai/latencytest/

**Comment by Holzhaus** [comment]
(by d00guan)
I tested updating to latest pulseaudio today, no ral improvement there. my previous version was one of the latest anyway.

I've tested running with 3 buffers before, and tried today but this don't give any improvement for me.

Tested setting nrofpacks, as recommended on the alsa link. No real improvement there.

I couldn't get latencytest to run. It's quite old and is missing things in my kernel.

I can indeed run Mixxx with ALSA directly and achieve latency around 5ms without much glitches, on both kernels. When they occur it's at track load. 
XRuns are definitely easier to reproduce with Jack.  With ck-preemptive kernel and jack Mixxx is quite stable at 11.6ms latency. Same here, glitches occur at track load if the occur. I can even start Firefox and Eclipse while playing without glitches.

**Comment by Holzhaus** [comment]
(by daniel-64studio)
I've done some more testing on a stock Lucid amd64 system using a fresh checkout from Mixxx bzr built with:

sudo scons prefix=/usr install tuned=1 debug=0 qdebug=0

So all debugging messages are turned off. The track load glitch is still audible with JACK at 256 x3, so debugging messages aren't the cause.

What I have discovered is that loading a FLAC file is likely to trigger an audible glitch, reported as an xrun by JACK, while loading a WAV does not. It doesn't matter if the track already playing is a FLAC or a WAV. So is it the overhead of decoding a FLAC that's the real cause of this?

Cheers! 

Daniel

**Comment by Holzhaus** [comment]
(by ywwg)
I think I can second RJ's conclusion above: any qdebug statement, even if it's piped to /dev/null, can cause major xruns.  For instance there's one in cachingreader.cpp that reports when there's a cache miss -- removing that debug statement eliminated my xruns.  So *any* qdebug in the audio processing thread has the potential to cause xruns.

Here's my recommendation:  create #ifdefs for all qdebug statements inside the audio threads so that when mixxx can be built with or without those debug statements triggered.  I think the default should be that debug=1 doesn't include those debug statements, but debug=2 enables them.  That way developers can still have useful debug information, but DJs can still use mix without horrific xruns on every track load.

I'm not sure how to use scons, but I can help putting in #ifdefs

**Comment by Holzhaus** [comment]
(by ywwg)
typo above: "so that when mixxx can be built"  should be "so that mixxx can be built"

**Comment by Holzhaus** [comment]
(by rryan)
Hi all,

While I don't think this bug is gone for good, there have been some big performance improvements in the 1.8 branch over the past week so Mixxx should be less CPU hungry. I wouldn't be surprised if that has improved this problem a little bit for some of you.

Thanks,
RJ

**Comment by Holzhaus** [comment]
(by daniel-64studio)
Hi Owen, see my comment #48 - I can reproduce this bug with debugging turned off completely if I load a FLAC file, but WAV files are OK. Are you playing WAVs or FLACs?

FLACs seem to decode faster than lossy files, but maybe there's a sudden hit on the CPU instead of a longer, lower load. Is it possible to lower the priority of the FLAC decoding process, like Mixxx already does for the analyzer thread, so that it doesn't compete with the audio callback thread?

**Comment by Holzhaus** [comment]
(by d00guan)
"create #ifdefs for all qdebug statements inside the audio threads so that when mixxx can be built with or without those debug statements triggered. I think the default should be that debug=1 doesn't include those debug statements, but debug=2 enables them. That way developers can still have useful debug information, but DJs can still use mix without horrific xruns on every track load."

Isn't it better to have a runtime debug flag? That way it's easier to get debug output from users if needed. The extra if statements should not affect the performance much.

**Comment by Holzhaus** [comment]
(by iltony)
Can confirm. Audio glitches while loading new tracks are very audibile. Tried both alsa and jack, two different soundcards (a Phonic Helix Board Firewire 18CH and a Hercules DJ Console Mk2), latency set to 23ms (quite high), tried RJ patch as well... no luck at all! BTW mixxx 1.7.2 was quite fine, so the question is: what changed in the audio engine between the two versions?
What I can see, is that now, when I load a track, cpu usage goes very very high (>100%), which in the past didn't appened. There's nothing wrong with it. But since we're in a real time context maybe if we can slow things down a bit (a sleep somewhere?)... But it's a really dirty approach... there should be something wrong about priorities.

**Comment by Holzhaus** [comment]
(by jkittsmiller2)
I'm also having this issue but with compressed audio only (mp3, ogg, flac, etc.) wav files are not affected, this may have something to do with the codec's. I believe it is probably more of a codec loading / unloading issue on Linux other than anything else, either the way in which Linux is handling them or the way in which Mixxx is accessing them. What does everyone else think?

**Comment by Holzhaus** [comment]
(by iltony)
I believe James Kittsmiller has it right. I did some detailed test, and I discoverd that that same issue is reproducible not only while loading a new track (with all the stuff such as bmp detection, waveform rendering and so on), but also everytime seeking through a new position of a (compressed) audio file.
Whenever seeking, to 0 (while loading) or to any position in the song, both audio streams stops for a frame or two before going on normally. The bug is more reproducibile while on a high cpu load, so for debugging porpouse it might a good idea to force such a situation and test it down. In order to verify this hypotesis about codecs, I tried to compile with ffmpeg support, but that didn't work because of compilation error. I also tried about forcing some memory unlock, but that didn't help either, so I'm propending towards codecs issue.

**Comment by Holzhaus** [comment]
(by wolfgang-winter-gmx)
I have the same "bug", audio xruns when loading a track at a windows System:

- Windows XP 32 bit
- Mixxx 1.8.0
- HDA-Intel Soundcard (internal)
- Asio4All
- Processor: Core 2 duo 6320

Without asio4all, with DirectSound, Sound running only very sloooooow. (not usable)

**Comment by Holzhaus** [comment]
(by iltony)
Oh yeah! Think I kinda got it. Here it is:

in file: mixxx/src/cachingreader.cpp
@line 514: freeAllChunks();

just comment the line, so that it looks like:

// freeAllChunks();

compile with scons and run again. Memory usage slowly rises while loading new pretty tracks, as expected, but, hey!, glitches are (mostly) gone. Looks like working. Obviously this ain't no production patch (and I'm not sure about it, I swam through the code and found this) but... does anybody want to give it try?

**Comment by Holzhaus** [comment]
(by rryan)
Hey iltony and everyone,

Thanks for your patience with us while we figure this bug out. It's a really tricky one. I'm testing a fix right now. I'll have a patch posted later tonight.

thanks,
RJ

**Comment by Holzhaus** [comment]
(by rryan)
Hi everyone,

Thanks to all your reports, I was able to track down what was causing this. There was a clear lock-up in the mixing engine which was triggered by loading a track while another was playing. 

If you are able to compile from source, please undo your edits and try the attached patch. Post here whether or not the patch worked for you. If it turns out that this patch fixes the problems people are having, we will release a Mixxx 1.8.1 with this and some other bug fixes applied very soon to get the fix out to everyone affected by this bug.

Thanks for your patience,
RJ Ryan

**Comment by Holzhaus** [comment]
(by rryan)
To Wolfgang Winter,

If your sound is running too slowly in general with ASIO, you probably need to increase your latency. 

RJ

**Comment by Holzhaus** [comment]
(by pegasus-renegadetech)
Patch tests good for me on a slow Linux system.

**Comment by Holzhaus** [comment]
(by iltony)
Mixxx works better now with the latest RJ patch (works like it did with mine). Glitches on audio while loading new tracks happen less frequently. Glitches on user interface are still very notable, but that's something which does not affect the quality of a live stage... Feels like there are still things which can be tuned here and there (since mixxx 1.7.2 achieved lower latencies), but the improvement is notable.

**Comment by Holzhaus** [comment]
(by rryan)
`teaser' on IRC also reports that this fixes the issue. Has anybody else tried it?

**Comment by Holzhaus** [comment]
(by jus)
As iltony stated there are still some audio issues, but the patch solves this particular issue for me with MacOSX. 
Tested with reference sound from /mixxx/src/test/soundFileFormats/

**Comment by Holzhaus** [comment]
(by d00guan)
I updated to latest from 1.8 branch and can confirm s a real improvement. The issue might be completely solved. Lately I've noticed longer a freeze if I've had one deck playing and one empty, and then loaded a track to the empty deck. This is completely gone now. You're my hero of the day RJ Ryan!

I can also confirm a huge over all performance improvement from using jackd2 over the old jackd in Ubuntu.

**Comment by Holzhaus** [comment]
(by gamegod)
Anders: I'm going to point the next asshole that comes along and blames
PortAudio for our performance problems and tells us to use the JACK API
natively at your comment.

Thanks for the ammo!

**Comment by Holzhaus** [comment]
(by rryan)
Since I've has many reports of this issue being fixed, I'm going to mark this fixed. This fix will be released in Mixxx 1.8.1 which we will release, along with some other bugfixes, very shortly.

Thanks to everyone who helped solve this with your useful debugging and feedback. People may experience xrun's on track load still, but this is probably not an inherent bug, rather a problem with having too-low latency or something similar.

**Comment by Holzhaus** [comment]
(by d00guan)
Albert: Please do! I'm not sure I can explain why though. Guess some improvements have been made to jack. I was quite puzzled after upgrading to Ubuntu 10.10, when I was able to run Mixxx almost glitch free with jack, on the generic kernel with CPU-scaling set to on-demand and before applying this patch.

I will also try to run jack with default time-out from now on to see if RJs patch helps with the disconnection problem.

**Comment by Holzhaus** [comment]
(by funkoolow)
hi,
i'm still experiencing the xruns on file load on ubuntu studio 10.10 and mixxx 1.8.1, running on eeepc900.
Maybe it's an hardware capability problem, but i'm not sure since everything runs very smooth except for that damn file load.
please let me know what to report to diagnose and thank you all for the patient support.

**Comment by Holzhaus** [comment]
(by rryan)
Hi funkoolow,

Could you please let us know which sound API you are using and also whether the xruns happen at any latency or just low-latency? Also, what is your CPU speed? 

Thanks,
RJ Ryan

**Comment by Holzhaus** [comment]
(by pegasus-renegadetech)
Also, are you using FLAC files? If so, you may be seeing bug #666052.

**Comment by Holzhaus** [comment]
(by iltony)
I'm deeply investigating the issue. To all those are experimenting sound glitches while loading a new track, could you please run this:

while true; do     if on_ac_power; then         nice /usr/bin/nvidia-settings -q all > /dev/null; cpufreq-selector -g performance;     else cpufreq-selector -g ondemand;     fi;     sleep 25; done

in a shell, and let it go while using mixxx? Do you see any improvement? It looks to me like there might still be some issue not strictly mixxx-related, but more than that originated from video driver preempting everything. Please let me know if you notice further improvement (notice: this applies only to those one who use an Nvidia graphic adapter).
